src/lib.rs

```diff
@@ -287,69 +272,622 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+            self.as_mut(),
+        );
+    }
+}
+
+impl<D: AsRef<[Digit]>> Mul<Digit> for &BigInt<D> {
     type Output = BigInt;
 
-    #[inline(always)]
-    fn add(self, rhs: u16) -> Self::Output {
-        self + Digit::from(rhs)
+    /// Computes `self * rhs` where `rhs` is a `Digit`.
+    ///
+    /// ## Constant time
+    /// Irrespective of it's inputs, this method takes `O(self.size())` steps.
+    ///
+    /// ## Zeroizing
+    /// This function might allocate some auxiliary buffer on the heap.
+    /// Before deallocating, it gets zeroed out.
+    ///
+    /// ## Complexity
+    /// The complexity is `O(self.size())`, which is asymptotically optimal.
+    fn mul(self, rhs: Digit) -> Self::Output {
+        self * &BigInt::from_digits(&[rhs])
     }
 }
 
-impl<D: AsRef<[Digit]>> Add<u32> for &BigInt<D> {
-    type Output = BigInt;
+impl<D: AsMut<[Digit]>> BigInt<D> {
+    /// Computes `a * b` where `b` is a digit, writes the result to `self`.
+    ///
+    /// It is a low-level function allowing you to manage allocations. Its safe analogue is
+    /// [`&a * b`](#impl-Mul%3CDigit%3E)
+    ///
+    /// ## Safety
+    /// Assuming that `a` has [size] `n`:
+    /// * `n > 0`
+    /// * [Size][size] of `self` must be exactly `n+1` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::multiplication(n, 1)`](ScratchSpace::multiplication) limbs
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n`.
+    pub unsafe fn mul_digit_unchecked(
+        &mut self,
+        a: &BigInt<impl AsRef<[Digit]>>,
+        b: Digit,
+        scratch_space: ScratchSpace,
+    ) {
+        self.mul_unchecked(a, &BigInt::from_digits(&[b]), scratch_space);
+    }
+}
 
-    #[inline(always)]
-    fn add(self, rhs: u32) -> Self::Output {
-        self + Digit::from(rhs)
+/// [Division](#impl-Div%3C%26%27_%20BigInt%3CD2%3E%3E) and [remainder](#impl-Rem%3C%26%27_%20BigInt%3CD2%3E%3E)
+/// operations require the dividend to start with a non-zero digit. This struct exists solely
+/// to emphasize this fact and is used as a type of dividends in these operations.
+/// Constant-time division is guaranteed relative to the size of unpadded dividends.
+pub struct Unpadded<D: AsRef<[Digit]>>(BigInt<D>);
+
+impl<D: AsRef<[Digit]>> BigInt<D> {
+    /// Remove all leading zeroes from a [`BigInt`](BigInt), converting it into an [unpadded `BigInt`](Unpadded).
+    pub fn as_unpadded(&self) -> Unpadded<Box<[Digit]>> {
+        let self_digits = self.as_ref();
+
+        // m is the length of self without leading zeroes
+        let m = ops::len_without_leading_zeroes(self_digits);
+
+        if m == 0 {
+            panic!("Unpadded BigInt should never be zero");
+        }
+
+        Unpadded(BigInt::from_digits(
+            self_digits[..m].to_vec().into_boxed_slice(),
```

Comment:
This line has a problem: `to_vec()` copies the slice. Not really sure how to fix this yet.